plutus-tx/src/Language/PlutusTx/Evaluation.hs

```diff
@@ -37,14 +37,12 @@ unsafeEvaluateCek
     => Program Name uni fun () -> EvaluationResult (Term Name uni fun ())
 unsafeEvaluateCek (Program _ _ t) = UPLC.unsafeEvaluateCekNoEmit defBuiltinsRuntime t
 
--- TODO: pretty sure we shouldn't need the unsafePerformIOs here, we should expose a pure interface even if it has IO hacks under the hood
-
 -- | Evaluate a program in the CEK machine with the usual string dynamic builtins and tracing, additionally
 -- returning the trace output.
 evaluateCekTrace
     :: (uni ~ DefaultUni, fun ~ DefaultFun)
     => Program Name uni fun ()
     -> ([String], CekExTally fun, Either (CekEvaluationException uni fun) (Term Name uni fun ()))
 evaluateCekTrace (Program _ _ t) =
-    case runCek defBuiltinsRuntime Counting True t of
-        (errOrRes, st, logs) -> (logs, toTally st, errOrRes)
+    case runCek defBuiltinsRuntime tallying True t of
+        (errOrRes, TallyingSt st _, logs) -> (logs, st, errOrRes)
```

Comment:
That was a bug: there's no point in extracting tallying info when running in the counting mode. Now that we don't have a single budgeting state type, it's not possible to type check this thing, which is how the bug was revealed. Yay proper typing.